src/Cake.Core/Scripting/Processors/ToolDirectiveProcessor.cs

```diff
@@ -0,0 +1,191 @@
+        /// <param name="environment">The environment.</param>
+        /// <param name="log">The log.</param>
+        /// <param name="nugetToolResolver">The NuGet tool resolver.</param>
+        public ToolDirectiveProcessor(IFileSystem fileSystem, ICakeEnvironment environment, ICakeLog log, INuGetToolResolver nugetToolResolver)
+            : base(environment)
+        {
+            if (fileSystem == null)
+            {
+                throw new ArgumentNullException("fileSystem");
+            }
+            if (environment == null)
+            {
+                throw new ArgumentNullException("environment");
+            }
+            if (log == null)
+            {
+                throw new ArgumentNullException("log");
+            }
+            if (nugetToolResolver == null)
+            {
+                throw new ArgumentNullException("nugetToolResolver");
+            }
+
+            _fileSystem = fileSystem;
+            _environment = environment;
+            _log = log;
+            _nugetToolResolver = nugetToolResolver;
+        }
+
+        /// <summary>
+        /// Processes the specified line.
+        /// </summary>
+        /// <param name="processor">The script processor.</param>
+        /// <param name="context">The script processor context.</param>
+        /// <param name="currentScriptPath">The current script path.</param>
+        /// <param name="line">The line to process.</param>
+        /// <returns>
+        ///   <c>true</c> if the processor handled the line; otherwise <c>false</c>.
+        /// </returns>
+        public override bool Process(IScriptProcessor processor, ScriptProcessorContext context, FilePath currentScriptPath, string line)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            var tokens = Split(line);
+            var directive = tokens.FirstOrDefault();
+            if (string.IsNullOrWhiteSpace(directive))
+            {
+                return false;
+            }
+
+            if (!directive.Equals("#tool", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Fetch the tool NuGet ID.
+            var toolId = tokens
+                .Select(value => value.UnQuote())
+                .Skip(1).FirstOrDefault();
+
+            if (string.IsNullOrWhiteSpace(toolId))
+            {
+                return false;
+            }
+
+            // Fetch optional NuGet source.
+            var source = tokens
+                .Skip(2)
+                .Select(value => value.UnQuote())
+                .FirstOrDefault();
+
+            // Get the directory path to Cake.
+            var applicationRoot = _environment.GetApplicationRoot();
+
+            // Get the tool directory.
+            var toolsRootDirectoryPath = applicationRoot
+                .Combine("..\\Tools")
```

Comment:
What happens if the user using the tool does not follow  the same conventions as we do, i.e. keeps `Cake.exe` in the root folder or at a location that's not expected. Will `../tools` still be a good path?